producer.go

```diff
@@ -201,7 +205,11 @@ func (w *Producer) sendCommandAsync(cmd *Command, doneChan chan *ProducerTransac
 
 func (w *Producer) connect() error {
 	w.guard.Lock()
-	defer w.guard.Unlock()
+	w.logGuard.RLock()
```

Comment:
let's wrap `logGuard` around a smaller critical section (without a defer) and store the `w.logger` in a local var for passing it along into the new connection instantiation